Added Sarvam TTS Websocket Implementation

pratham-sarvam opened the pull request:
Added TTS websocket implementation for Sarvam.

vinayak-sarvam commented:
Hi @markbackman, Can you please review and provide your feedback on this PR.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
             yield TTSStoppedFrame()
+
+
+class SarvamWebsocketTTSService(WebsocketTTSService):
```
The convention is for the websocket service to be called `ServiceTTSService` and the HTTP service to be called `ServiceHttpTTSService`. So, let's update the HTTP implementation to `SarvamHttpTTSService`.

Also, this will need to be a different base class—`InterruptibleTTSService`. (I know, it's confusing! We're refactoring the base classes.)

```suggestion
class SarvamTTSService(InterruptibleTTSService):
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+
+
+class SarvamWebsocketTTSService(WebsocketTTSService):
+    """Minimalist TTS service for Sarvam AI WebSocket endpoint using Pipecat base class.
```
These docstrings are used in auto-generating docs, so you might want something a little more complete.

Check out the docstring from the current service as an example.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -6,23 +6,31 @@
 from typing import AsyncGenerator, Optional
 
 import aiohttp
+import websockets
```
This needs to be in a try/except. Also, we support the asyncio implementation now, so you need to implement websockets slightly differently. I'll comment throughout.

```
try:
    from websockets.asyncio.client import connect as websocket_connect
    from websockets.protocol import State
except ModuleNotFoundError as e:
    logger.error(f"Exception: {e}")
    logger.error("In order to use Sarvam, you need to `pip install pipecat-ai[sarvam]`.")
    raise Exception(f"Missing module: {e}")
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        if isinstance(frame, (TTSStoppedFrame, StartInterruptionFrame)):
+            self._started = False
+
+    async def _flush_and_reconnect(self):
```
This is unused. Can we remove?

pratham-sarvam replied:
This is not removed as of now, might need it in change voice function, will remove after implementing all the changes.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            subprotocols = [f"api-subscription-key.{self._api_key}"]
+            self._websocket = await websockets.connect(
+                self._websocket_url,
+                subprotocols=subprotocols,
```
Should we just use `additional_headers` (ws asyncio replacement for extra_headers) to set the X-API-Key?

pratham-sarvam replied:
Addressed in a PR

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            self._websocket = await websockets.connect(
+                self._websocket_url,
+                subprotocols=subprotocols,
+                ping_interval=None,  # We'll handle pings manually
```
Why are these pings and timeouts needed?

pratham-sarvam replied:
Our websocket automatically closes in case there is no incoming message for a timeout period (currently 1 min) so in case we want persistent connection we need to send ping message.

markbackman replied:
Does the keepalive messages accomplish that? You're sending messages every 10 seconds.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        self.set_model_name(model)
+        self.set_voice(speaker)
+        # Configuration parameters
+        self._config = {
```
nit: we've called this `_settings` in other TTS services.

pratham-sarvam replied:
Addressed in a PR

vinayak-sarvam commented:
Hi @markbackman, Appreciate your prompt response, thanks for the detailed feedback. Our team will look into this.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        # Initialize parent class first
+        super().__init__(
+            push_stop_frames=True,
+            sample_rate=speech_sample_rate,
```
```suggestion
            sample_rate=sample_rate,
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        Args:
+            frame: The start frame containing initialization parameters.
+        """
+        await super().start(frame)
```
Assuming the naming changes I'm recommending, we initialize the sample_rate here.

```suggestion
        await super().start(frame)
        self._settings["sample_rate"] = self.sample_rate
```

This sets the value from the TTSService base class, which is handling the StartFrame and exposing a `sample_rate` property.

pratham-sarvam replied:
Addressed in a PR

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+
+                    await self.push_frame(ErrorFrame(f"TTS Error: {error_msg}"))
+
+                elif msg.get("type") == "pong":
```
Is this needed?

pratham-sarvam replied:
This might not be needed. Will remove this after the final changes

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+                    await self.start_ttfb_metrics()
+                    yield TTSStartedFrame()
+                    self._started = True
+                    self._cumulative_time = 0
```
Remove: unused. This would be used for word/timestamp pairs.

pratham-sarvam replied:
Which variable are we talking about here exactly? What is the usecase behind each of them?

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            except Exception as e:
+                logger.error(f"{self} error sending message: {e}")
+                yield TTSStoppedFrame()
+                logger.debug("error wala disconnect")
```
What is `wala`?

pratham-sarvam replied:
Addressed in a PR

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        """
+        return True
+
+    def language_to_service_language(self, language: Language) -> Optional[str]:
```
You might want to add a method to change voices. I'm assuming we'd have to disconnect the websocket to accomplish that, right? (Assuming suggested names)

```
    async def _update_settings(self, settings: Mapping[str, Any]):
        """Update service settings and reconnect if voice changed."""
        prev_voice = self._voice_id
        await super()._update_settings(settings)
        if not prev_voice == self._voice_id:
            logger.info(f"Switching TTS voice to: [{self._voice_id}]")
            await self._disconnect()
            await self._connect()
```

markbackman replied:
This uses the base class to handle the frame and update the value. You can test this by pushing a `TTSUpdateSettingsFrame`:

In your example:
```
    @transport.event_handler("on_client_connected")
    async def on_client_connected(transport, client):
        logger.info(f"Client connected")
        # Kick off the conversation.
        messages.append({"role": "system", "content": "Please introduce yourself to the user."})
        await task.queue_frames([context_aggregator.user().get_context_frame()])
        # Wait 15 seconds and then change the voice.
        await asyncio.sleep(15)
        await task.queue_frame(TTSUpdateSettingsFrame(settings={"voice_id": "new-voice-id"}))
```

pratham-sarvam replied:
Addressed in a PR
I have added this logic here but somehow the disconnect isn't happening, can you help guide what could be going wrong.

pratham-sarvam replied:
Sorry, this is an old comment, we addressed that issue, our websocket doesn't need to disconnect to change voices, we can send config message and it works as is. There was some small issue we fixed it (in the example file)

pratham-sarvam commented on examples/foundational/07z-interruptible-sarvam.py:
```diff
@@ -18,7 +18,7 @@
 from pipecat.services.deepgram.stt import DeepgramSTTService
 from pipecat.services.openai.llm import OpenAILLMService
-from pipecat.services.sarvam.tts import SarvamTTSService
+from pipecat.services.sarvam.tts import SarvamWebsocketTTSService
```
Addressed in a PR

pratham-sarvam commented on examples/foundational/07z-interruptible-sarvam.py:
```diff
@@ -56,10 +56,11 @@ async def run_example(transport: BaseTransport, _: argparse.Namespace, handle_si
     async with aiohttp.ClientSession() as session:
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        api_key: str,
+        model: str = "bulbul:v2",
+        target_language_code: str = "hi-IN",
+        pitch: float = 0.0,
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        URL based on the chosen model, and initializes state flags before connecting.
+        """
+        # Initialize parent class first
+        super().__init__(
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        )
+
+        # WebSocket endpoint URL
+        self._websocket_url = f"wss://api.sarvam.ai/text-to-speech/ws?model={model}"
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+    async def _connect_websocket(self):
+        """Establish WebSocket connection to Sarvam API."""
+        try:
+            if self._websocket and not self._websocket.closed:
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+
+            logger.debug("Connecting to Sarvam TTS Websocket")
+            subprotocols = [f"api-subscription-key.{self._api_key}"]
+            self._websocket = await websockets.connect(
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        pace: float = 1.0,
+        speaker: str = "anushka",
+        loudness: float = 1.0,
+        speech_sample_rate: int = 24000,
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            "pace": pace,
+            "speaker": speaker,
+            "loudness": loudness,
+            "speech_sample_rate": speech_sample_rate,
```
Addressed in a PR

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+
+    async def _receive_messages(self):
+        """Receive and process messages from Sarvam WebSocket."""
+        async for message in WatchdogAsyncIterator(self._websocket, manager=self.task_manager):
```
Addressed in a PR

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +261,406 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            aggregate_sentences: Whether to merge multiple sentences into one audio chunk (default True).
+            sample_rate: Desired sample rate for the output audio in Hz (overrides default if set).
+            params: Optional input parameters to override global configuration.
+            target_language_code: BCP-47 code of the target language (default “hi-IN”).
```
Remove the rest of the args from `target_language_code` through `output_audio_bitrate` (keeping **kwargs). These are now defined as `Parameters:` in `InputParams`.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +248,327 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        }
+        # self._websocket = None
+        self._started = False
+        self._cumulative_time = 0
```
Still unused. Remove.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +261,406 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+                    await self.stop_ttfb_metrics()
+                    audio = base64.b64decode(msg["data"]["audio"])
+                    frame = TTSAudioRawFrame(audio, self.sample_rate, 1)
+                    logger.debug("pushing audio frames from tts to output")
```
Remove this log line.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -6,23 +6,40 @@
     ErrorFrame,
     Frame,
     StartFrame,
+    StartInterruptionFrame,
```
Unused import. Remove.
```suggestion
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +261,406 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        Args:
+            frame: The cancel frame.
+        """
+        logger.debug("getting cancel frame")
```
Remove logline.

```suggestion
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -6,23 +6,40 @@
 
 import aiohttp
+
+try:
```
Move this try/except block to the bottom of the imports.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +261,406 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+        """Establish WebSocket connection to Sarvam API."""
+        try:
+            if self._websocket and self._websocket.state is State.OPEN:
+                logger.debug("Webscoket already connected")
```
Remove logline.

```suggestion
```

markbackman commented:
Looks really close! Just a little more clean up. Also, please add a changelog entry.

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -97,7 +112,7 @@ def __init__(
         voice_id: str = "anushka",
         model: str = "bulbul:v2",
-        aiohttp_session: aiohttp.ClientSession,
+        aiohttp_session: Optional[aiohttp.ClientSession] = None,
```
We are inclined to keep it optional as well, since not a lot of other implementations are following this structure of passing session into a Frame processor, so we feel it's better to keep it uniform. Moved down

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -324,6 +340,7 @@ def __init__(
         model: str = "bulbul:v2",
         voice_id: str = "anushka",
         url: str = "wss://api.sarvam.ai/text-to-speech/ws",
+        aiohttp_session: Optional[aiohttp.ClientSession] = None,
```
This is specifically put in to ensure backward compatibility, to ensure that if people are using our old code which refers to SarvamTTSService as the http implementation as opposed to the new implementation which now refers to websocket implementation. So, we want the old code to still work.

pratham-sarvam commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +277,371 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            msg = {"type": "flush"}
+            await self._websocket.send(json.dumps(msg))
+
+    async def process_frame(self, frame: Frame, direction: FrameDirection):
```
This is needed for our websocket class to send any pending response. We send this flush signal at the end of LLM response as seen in wordTTSService implementation :
```
        elif isinstance(frame, (LLMFullResponseEndFrame, EndFrame)):
            await self.flush_audio()
```
Do let us know if we should call flush_audio someway else since in TTS service I only see flush audio being called at TTSSpeak Frame received.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -97,9 +112,9 @@ def __init__(
-        aiohttp_session: aiohttp.ClientSession,
         base_url: str = "https://api.sarvam.ai",
         sample_rate: Optional[int] = None,
+        aiohttp_session: Optional[aiohttp.ClientSession] = None,
```
Also, after the change, resort the inputs—keep requires and optionals together.

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -109,24 +124,26 @@ def __init__(
         self._api_key = api_key
         self._base_url = base_url
-        self._session = aiohttp_session
+        self._session = aiohttp_session or aiohttp.ClientSession()
```
Required after the change I suggested.

```suggestion
        self._session = aiohttp_session
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +277,371 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            model: Identifier of the Sarvam speech model (default "bulbul:v2").
+            voice_id: Voice identifier for synthesis (default "anushka").
+            url: WebSocket URL for connecting to the TTS backend (default production URL).
+            aiohttp_session: Optional shared aiohttp session. To maintain backward compatibility.
```
```suggestion
            aiohttp_session: Optional shared aiohttp session. To maintain backward compatibility.
            
                .. deprecated:: 0.0.81
                    aiohttp_session is no longer used. This parameter will be removed in a future version.
```

markbackman commented on src/pipecat/services/sarvam/tts.py:
```diff
@@ -240,3 +277,371 @@ async def run_tts(self, text: str) -> AsyncGenerator[Frame, None]:
+            **kwargs,
+        )
+        params = params or SarvamTTSService.InputParams()
+
```
Add:

```
        if aiohttp_session is not None:
            warnings.warn(
                "The 'aiohttp_session' parameter is deprecated and will be removed in a future version. "
                DeprecationWarning,
                stacklevel=2,
            )
```

pratham-sarvam commented:
You can look at this log line : `Generating TTS: [text]` and see that the last segment is sent but is not played back in audio, which is the issue of not flushing at LLM end response.

pratham-sarvam commented:
@markbackman We cannot merge this yet as when I was testing interruptions, I saw cases where the TTS just stops giving output and doesn't respond post that. Can we have a look at why that is happening? Is it due to something which is happening in the base classes and we are not handling in our implementation or is it some other bug we are missing? Happy to schedule a short call as well in case that is required.

markbackman commented:
> @markbackman We cannot merge this yet as when I was testing interruptions, I saw cases where the TTS just stops giving output and doesn't respond post that. Can we have a look at why that is happening? Is it due to something which is happening in the base classes and we are not handling in our implementation or is it some other bug we are missing? Happy to schedule a short call as well in case that is required.

I have not seen this. A few things:
1. Can you rebase on `main`?
2. What python version are you running?
3. What's the repro case? I tested interruptions and did not see this happening.

pratham-sarvam commented:
> > @markbackman We cannot merge this yet as when I was testing interruptions, I saw cases where the TTS just stops giving output and doesn't respond post that. Can we have a look at why that is happening? Is it due to something which is happening in the base classes and we are not handling in our implementation or is it some other bug we are missing? Happy to schedule a short call as well in case that is required.
> 
> I have not seen this. A few things:
> 
> 1. Can you rebase on `main`?
> 2. What python version are you running?
> 3. What's the repro case? I tested interruptions and did not see this happening.

1. I did now.
2. Python 3.11.5
3. It is the case where I keep interrupting the bot and this is the error I get after some time, I have added a fix can you please have a look. This error doesn't repro after this fix :

 ```
Received an unexpected media stream error while reading the audio.
2025-08-20 09:29:45.093 | DEBUG    | pipecat.transports.network.webrtc_connection:on_iceconnectionstatechange:289 - ICE connection state is closed, connection is closed
2025-08-20 09:29:45.093 | DEBUG    | pipecat.transports.network.webrtc_connection:_handle_new_connection_state:454 - Connection state changed to: closed
2025-08-20 09:29:45.093 | INFO     | pipecat.runner.run:handle_disconnected:221 - Discarding peer connection for pc_id: SmallWebRTCConnection#0
2025-08-20 09:29:45.093 | DEBUG    | pipecat.transports.network.small_webrtc:on_closed:247 - Client connection closed.
2025-08-20 09:29:45.094 | INFO     | __main__:on_client_disconnected:110 - Client disconnected
2025-08-20 09:29:45.094 | DEBUG    | pipecat.pipeline.task:_cancel:492 - Cancelling pipeline task PipelineTask#0
2025-08-20 09:29:45.095 | DEBUG    | pipecat.services.deepgram.stt:_disconnect:239 - Disconnecting from Deepgram
2025-08-20 09:29:46.449 | DEBUG    | pipecat.services.sarvam.tts:_disconnect_websocket:537 - Disconnecting from Sarvam
^CINFO:     Shutting down
INFO:     Waiting for background tasks to complete. (CTRL+C to force quit)
^CINFO:     Finished server process [88216]
ERROR:    Traceback (most recent call last):
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1512, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1505, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1379, in uvloop.loop.Loop.run_forever
  File "uvloop/loop.pyx", line 557, in uvloop.loop.Loop._run
  File "uvloop/loop.pyx", line 476, in uvloop.loop.Loop._on_idle
  File "uvloop/cbhandles.pyx", line 83, in uvloop.loop.Handle._run
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/uvicorn/server.py", line 70, in serve
    with self.capture_signals():
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/uvicorn/server.py", line 331, in capture_signals
    signal.raise_signal(captured_signal)
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 157, in _on_sigint
    raise KeyboardInterrupt()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/starlette/routing.py", line 701, in lifespan
    await receive()
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/uvicorn/lifespan/on.py", line 137, in receive
    return await self.receive_queue.get()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/queues.py", line 158, in get
    await getter
asyncio.exceptions.CancelledError

2025-08-20 09:29:51.579 | WARNING  | pipecat.pipeline.task:_print_dangling_tasks:834 - Dangling tasks detected: ['SmallWebRTCOutputTransport#0::_clock_task_handler', 'SmallWebRTCOutputTransport#0::_audio_task_handler']
2025-08-20 09:29:51.579 | DEBUG    | pipecat.pipeline.runner:run:88 - Runner PipelineRunner#0 finished running PipelineTask#0
ERROR:asyncio:an error occurred during closing of asynchronous generator <async_generator object BaseOutputTransport.MediaSender._next_frame.<locals>.without_mixer at 0x163117140>
asyncgen: <async_generator object BaseOutputTransport.MediaSender._next_frame.<locals>.without_mixer at 0x163117140>
Traceback (most recent call last):
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1512, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1505, in uvloop.loop.Loop.run_until_complete
  File "uvloop/loop.pyx", line 1379, in uvloop.loop.Loop.run_forever
  File "uvloop/loop.pyx", line 557, in uvloop.loop.Loop._run
  File "uvloop/loop.pyx", line 476, in uvloop.loop.Loop._on_idle
  File "uvloop/cbhandles.pyx", line 83, in uvloop.loop.Handle._run
  File "uvloop/cbhandles.pyx", line 63, in uvloop.loop.Handle._run
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/uvicorn/server.py", line 70, in serve
    with self.capture_signals():
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/contextlib.py", line 144, in __exit__
    next(self.gen)
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/venv/lib/python3.11/site-packages/uvicorn/server.py", line 331, in capture_signals
    signal.raise_signal(captured_signal)
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/runners.py", line 157, in _on_sigint
    raise KeyboardInterrupt()
KeyboardInterrupt

During handling of the above exception, another exception occurred:

RuntimeError: aclose(): asynchronous generator is already running
Exception ignored in: <coroutine object BaseOutputTransport.MediaSender._clock_task_handler at 0x133291c40>
Traceback (most recent call last):
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/src/pipecat/transports/base_output.py", line 846, in _clock_task_handler
    timestamp, _, frame = await self._clock_queue.get()
                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/prathamgupta/Documents/Sarvam/pipecat/src/pipecat/utils/asyncio/watchdog_priority_queue.py", line 84, in get
    get_result = await super().get()
                 ^^^^^^^^^^^^^^^^^^^
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/queues.py", line 160, in get
    getter.cancel()  # Just in case getter is not done yet.
    ^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1281, in uvloop.loop.Loop.call_soon
  File "uvloop/loop.pyx", line 669, in uvloop.loop.Loop._call_soon
  File "uvloop/loop.pyx", line 678, in uvloop.loop.Loop._call_soon_handle
  File "uvloop/loop.pyx", line 673, in uvloop.loop.Loop._append_ready_handle
  File "uvloop/loop.pyx", line 705, in uvloop.loop.Loop._check_closed
RuntimeError: Event loop is closed
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending name='SmallWebRTCOutputTransport#0::_clock_task_handler' coro=<TaskManager.create_task.<locals>.run_coroutine() done, defined at /Users/prathamgupta/Documents/Sarvam/pipecat/src/pipecat/utils/asyncio/task_manager.py:234> wait_for=<Future cancelled> cb=[TaskManager._task_done_handler()]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending name='SmallWebRTCOutputTransport#0::_audio_task_handler' coro=<TaskManager.create_task.<locals>.run_coroutine() done, defined at /Users/prathamgupta/Documents/Sarvam/pipecat/src/pipecat/utils/asyncio/task_manager.py:234> wait_for=<Future pending cb=[Task.task_wakeup()]> cb=[TaskManager._task_done_handler()]>
ERROR:asyncio:Task was destroyed but it is pending!
task: <Task pending name='Task-32140' coro=<Queue.get() running at /Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/queues.py:160> wait_for=<Future cancelled> cb=[_release_waiter(<Future pendi...ask_wakeup()]>)() at /Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/tasks.py:421]>
Exception ignored in: <coroutine object Queue.get at 0x163117e60>
Traceback (most recent call last):
  File "/Users/prathamgupta/.pyenv/versions/3.11.5/lib/python3.11/asyncio/queues.py", line 160, in get
    getter.cancel()  # Just in case getter is not done yet.
    ^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1281, in uvloop.loop.Loop.call_soon
  File "uvloop/loop.pyx", line 669, in uvloop.loop.Loop._call_soon
  File "uvloop/loop.pyx", line 678, in uvloop.loop.Loop._call_soon_handle
  File "uvloop/loop.pyx", line 673, in uvloop.loop.Loop._append_ready_handle
  File "uvloop/loop.pyx", line 705, in uvloop.loop.Loop._check_closed
RuntimeError: Event loop is closed
```

There are dangling tasks due to race conditions and I have addressed those, please check

filipi87 reviewed (approved):
Looking at the code, it looks great. I’ve also tested both examples, and they are working nicely.

I’ll approve and merge this. I’ll add the changelog in a follow-up PR, so this one doesn’t stay blocked.

Thank you for doing this!

vinayak-sarvam commented:
Thank you @markbackman and @filipi87 for your reviews, suggestions and helping on this on. We really appreciate it.

pratham-sarvam commented:
Thanks @markbackman and @filipi87 for prompt feedbacks and testing. It was really helpful as well.
